python/ray/serve/batching.py

```diff
@@ -172,45 +190,82 @@ async def _consume_func_generator(
         FINISHED_TOKEN = None
 
         try:
-            futures = initial_futures
+            futures = deque(initial_futures)
+            assert len(futures) == input_batch_length
+
             async for results in func_generator:
                 self._validate_results(results, input_batch_length)
-                next_futures = []
-                for result, future in zip(results, futures):
+                for idx in range(input_batch_length):
+                    result, future = results[idx], futures[0]
+
                     if future is FINISHED_TOKEN:
                         # This caller has already terminated.
-                        next_futures.append(FINISHED_TOKEN)
-                    elif result in [StopIteration, StopAsyncIteration]:
+                        futures.append(FINISHED_TOKEN)
```

Comment:
I still don't understand what we need FINISHED_TOKEN for (saw your response on the outdated thread)?

We're adding it to the queue but not really handling it in any meaningful way as far as i can tell

Reply:
We discussed offline; the `FINISHED_TOKEN` is necessary for padding.